### ISSUE TITLE
NLL: spurious error emitted on body of (erroneous) const fn body

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#55825]
Spawned off of #54528 (specificially [this card](https://github.com/rust-lang/rust/projects/10#card-14187258) in the project).

Consider the following code:

```rust
const fn no_dyn_trait_ret() -> &'static dyn std::fmt::Debug { &() }

fn main() { }
```

With AST-borrowck, you get ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2015&gist=72bb00525112d521bcd693ae453a9cf0)):

```
error: trait bounds other than `Sized` on const fn parameters are unstable
 --> src/main.rs:1:32
  |
1 | const fn no_dyn_trait_ret() -> &'static dyn std::fmt::Debug { &() }
  |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error
```

With NLL (here using migration a la 2018 edition), you get ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d8f02adc3bd04d538ce7cd40143b93b3)):

```
error: trait bounds other than `Sized` on const fn parameters are unstable
 --> src/main.rs:1:32
  |
1 | const fn no_dyn_trait_ret() -> &'static dyn std::fmt::Debug { &() }
  |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^

warning[E0515]: cannot return reference to temporary value
 --> src/main.rs:1:63
  |
1 | const fn no_dyn_trait_ret() -> &'static dyn std::fmt::Debug { &() }
  |                                                               ^--
  |                                                               ||
  |                                                               |temporary value created here
  |                                                               returns a reference to data owned by the current function
  |
  = warning: This error has been downgraded to a warning for backwards compatibility with previous releases.
          It represents potential unsoundness in your code.
          This warning will become a hard error in the future.

error: aborting due to previous error
```

I have not managed to replicate the above borrow-check error on code that does not exhibit some other `const fn` error.

And the problem goes away if you opt into `#![feature(const_fn)]` ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1e6085697b0092eeda3f612659f78bf3)).

So, this is really a *very* minor diagnostic issue.

**Comment by RalfJung** [comment]
Cc @oli-obk might be something in (min) const qualification?

**Comment by oli-obk** [comment]
TLDR: The issue with this is that we have a bunch of separate checks relating to const things, and most of these are spaghetti entangled into one. @eddyb and I are working towards fixing that, but it's nontrivial to say the least.

Details:

We have the following checks that we need to run on MIR (which relate to const):

* min_const_fn compliance
* const_fn compliance
* promotion of runtime values to an anonymous static/constant
* const compliance (valid array length, const item initializer or enum discriminant)
* static compliance (valid static item initializer)
* no interior mutability checks
* no drop in const checks

The current implementation has an early abort for const fns where if the `const_fn` feature gate is not active, we first run `min_const_fn`, and if that fails we don't run the entangled other 4 passes. If we ran the other passes in case of `min_const_fn` failure, you'd get a lot of error duplication.

I don't think we have a tracking issue for the refactoring separating all these passes into distinct dataflow passes, but there's ongoing work.

**Comment by Aaron1011** [comment]
This now produces the same output both with and without `#![feature(nll)]`.
